designs/1951-required-fields-v2.md

```diff
@@ -0,0 +1,352 @@
+All derived fields use regular, non-optional, fields.
+If a field should only be added conditionally, one currently has to revert to rather verbose methods, such as
+```
+a?: int
+b?: int
+if a != _|_ && b != _|_ {
+  c: a + b
+}
+```
+Case 3 is common (see https://github.com/cue-lang/cue/issues/1232).
+Having something much more compact would be nice and some thoughts for a proposal for that are in progress.
+
+### Policy
+Policy is like schema but tends to have more derived constraints.
+A key difference with schema is that policy is often not defined by the team or organization
+that defined the original schema.
+A policy typically just adds constraints to an existing schema.
+An example use case typical to policy is making a field required that was previously not required.
+Another is making constraints more specific, without making them concrete or implicitly defined.
+
+```
+#Def: {
+    name?: string
+    speed?: number
+}
+#DefPolicy: #Def & {
+    name?:  strings.MinRunes(1)
+    speed?: >100
+}
+```
+# Proposal
+CUE currently has the “optional field constraint”, denoted `foo?: value`.
+
+We propose adding a “required field constraint”, denoted `foo!: value`, which is like `foo?: value`,
+but requires a regular value be provided for `foo` (a field `foo:` without  `!:` or ?:`).
+
+We refer to optional field constraints and required field constraints collectively as “field constraints”.
+
+As a general rule, all data and data templating would be defined with regular fields,
+whereas schemas would be defined in terms of field constraints.
+Of course, CUE being CUE, mixing these two fields is allowed:
+this rule is not a restriction but suggested as a matter of style and proper usage.
+
+## Semantics
+Informally, requiredness `!` is a property of a field that is unified independently of a field value.
+
+The following holds:
+```
+{a?: x} > {a!: x} > {a: x}
+```
+for any `x`, such as `int` or `1`.
+
+For instance:
+```
+{foo!: 3} & {foo: 3}  	→  {foo: 3}
+{foo!: 3} & {foo: int}   	→  {foo: 3} // X
+{foo!: 3} & {foo: <=4}	   	→  {foo: 3}  // X
+
+{foo!: int} & {foo: int} 	  	→  {foo: int}
+{foo!: int} & {foo?: <1}   	→  {foo!: <1}
+{foo!: int} & {foo: <=3}   	→  {foo: <=3}
+{foo!: int} & {foo: 3}  		→  {foo: 3}
+```
+Note the subtle, but important difference between this proposal and #822: `foo!: value`
+only requires that a field be specified; it does not require it to be concrete.
+As we are no longer proposing a change to the semantics of `x: string`, though, the net effect is roughly the same.
+
+As a consequence, though, it is now possible to create a concrete value without
+having the user actually provide it (see the line marked `X` in the above example).
+So this loses the strictness of the original proposal.
+In practice, this likely is not much of an impediment: JSON and YAML cannot specify such fields,
+and thus this would not limit the ability to validate such files: they still need to specify
+the actual concrete value to pass.
```

Comment:
It might be good to say that the proposal is technically complete at this point?

```suggestion
the actual concrete value to pass.

That completes the proposal.
```